resources/windows/msi/main.wxs.erb

```diff
@@ -0,0 +1,781 @@
+      <Custom Action='SaveCmdLinePuppetAgentEnvironment' Before='AppSearch' />
+      <Custom Action='SetFromCmdLinePuppetAgentEnvironment' After='AppSearch'>
+        CMDLINE_PUPPET_AGENT_ENVIRONMENT
+      </Custom>
+      <Custom Action='SetDefaultPuppetAgentEnvironment' After='AppSearch'>
+        PUPPET_AGENT_ENVIRONMENT=""
+      </Custom>
+       <!-- PUPPET_AGENT_CERTNAME -->
+      <Custom Action='SaveCmdLinePuppetAgentCertname' Before='AppSearch' />
+      <Custom Action='SetFromCmdLinePuppetAgentCertname' After='AppSearch'>
+        CMDLINE_PUPPET_AGENT_CERTNAME
+      </Custom>
+      <!-- PUPPET_CA_SERVER -->
+      <Custom Action='SaveCmdLinePuppetCaServer' Before='AppSearch' />
+      <Custom Action='SetFromCmdLinePuppetCaServer' After='AppSearch'>
+        CMDLINE_PUPPET_CA_SERVER
+      </Custom>
+      <!-- PUPPET_AGENT_START_MODE -->
+      <Custom Action='SaveCmdLinePuppetAgentStartupMode' Before='AppSearch' />
+      <Custom Action='SetFromCmdLinePuppetAgentStartupMode' After='AppSearch'>
+        CMDLINE_PUPPET_AGENT_STARTUP_MODE
+      </Custom>
+      <!-- PUPPET_AGENT_ACCOUNT_DOMAIN -->
+      <Custom Action='SetDomainToLocalComputer' After='AppSearch'>
+        PUPPET_AGENT_ACCOUNT_DOMAIN = "."
+      </Custom>
+      <Custom Action='SetDomainToNtAuthority' After='SetDomainToLocalComputer'>
+        (PUPPET_AGENT_ACCOUNT_USER ~= "LocalService") OR (PUPPET_AGENT_ACCOUNT_USER ~= "NetworkService")
+      </Custom>
+      <?if $(var.Platform) = x86 ?>
+      <Custom Action='Remove64BitPath_SetProp' After='CostFinalize' />
+      <Custom Action='Remove64BitProgramFiles_SetProp' After='CostFinalize' />
+      <Custom Action='Remove64BitProgramFiles' After='InstallFiles'>
+        <![CDATA[VersionNT64 >= 100 AND $(var.Win64) = no AND NOT (&$(var.OurProductNameWord)Runtime = 2)]]>
+      </Custom>
+      <Custom Action='Remove64BitPath' After='InstallFiles'>
+        <![CDATA[VersionNT64 >= 100 AND $(var.Win64) = no AND NOT (&$(var.OurProductNameWord)Runtime = 2)]]>
+      </Custom>
+      <?endif?>
+      <!-- Make sure pxp-agent service is shutdown before we validate the install -->
+      <!-- Only executed if pxp-agent is already installed (i.e. an Uninstall or upgrade - check for repair also) -->
+      <Custom Action='ShutdownPxpAgentService_SetProp' Before='ShutdownPxpAgentService'>
+        REMOVE="ALL" OR WIX_UPGRADE_DETECTED or REINSTALL
+      </Custom>
+      <Custom Action='ShutdownPxpAgentService' Before='InstallValidate'>
+        REMOVE="ALL" OR WIX_UPGRADE_DETECTED or REINSTALL
+      </Custom>
+      <!--
+      WIX_UPGRADE_DETECTED
+      -->
+    </InstallExecuteSequence>
+
+    <!--
+      The Features to install.  We're using ComponentGroupRef instead of ComponentRef since we're
+      generating the ruby wxs file using heat
+    -->
+    <Feature Id="$(var.OurProductNameWord)Runtime" Title="$(var.OurProductName) Runtime" Level="1">
+      <ComponentRef Id="RegistryEntries" />
```

Comment:
If we name the ComponentGroup purposefully in the heat generation, we can most likely reduce this list of comonentrefs to one.